test/evp_extra_test.c

```diff
@@ -4647,6 +4657,80 @@ const OPTIONS *test_get_options(void)
     return options;
 }
 
+/* Test that trying to sign with a public key errors out gracefully */
+static int test_ecx_not_private_key(int tst)
+{
+#ifdef OPENSSL_NO_EC
+    return 1;
+#else
+    EVP_PKEY *pkey = NULL;
+
+    const unsigned char msg[] = { 0x00, 0x01, 0x02, 0x03 };
+    int testresult = 0;
+    EVP_MD_CTX *ctx = NULL;
+    unsigned char *mac = NULL;
+    size_t maclen = 0;
+    int ret;
+    unsigned char *pubkey;
+    size_t pubkeylen;
+    OSSL_PROVIDER* deflprov;
+
+
+    switch (keys[tst].type) {
+        case NID_X25519:
+        case NID_X448:
+            /* Test does not apply, skip */
+            return 1;
+    }
+
+    /* Check if this algorithm supports public keys */
+    if (keys[tst].pub == NULL)
+        return 1;
+
+    deflprov = OSSL_PROVIDER_load(NULL, "default");
+
+    if (!TEST_ptr(deflprov))
+        return 0;
+
+    pubkey = (unsigned char *)keys[tst].pub,
+    pubkeylen = strlen(keys[tst].pub);
+
+    pkey = EVP_PKEY_new_raw_public_key(keys[tst].type, NULL, pubkey, pubkeylen);
+
+    if (!TEST_ptr(pkey))
+        goto err;
+
+    if (!TEST_ptr(ctx = EVP_MD_CTX_new()))
+        goto err;
+
+    ret = EVP_DigestSignInit(ctx, NULL, NULL, NULL,
+                             pkey);
+
+    if (!TEST_true(ret))
+        goto err;
+
+    if (!TEST_true(EVP_DigestSign(ctx, NULL, &maclen, msg, sizeof(msg))))
+        goto err;
```

Comment:
In theory, the provider could fail the operation already at the EVP_DigestSignInit() or the EVP_DigestSign() call with NULL signature buffer. I am not sure it is good idea to cement this behavior by the test. Failure of either of these three calls should be accepted as correct.

Reply:
I agree, theoritically those two functions would also fail. But I'm bit concerned if some other failure in those two functions will mask the test, as we don't actually test the error message. How do you think?

Reply:
You could call ERR_peek_error() and test that there is the appropriate error reason code on the stack.